code/datums/components/trader/trader.dm

```diff
@@ -0,0 +1,413 @@
+	trader.say(trader_data.return_trader_phrase(BUY_PHRASE))
+
+///Calculates the value of money in the hand of the buyer and spends it if it's sufficient
+/datum/component/trader/proc/spend_buyer_offhand_money(mob/customer, the_cost)
+	var/value = 0
+	var/obj/item/holochip/cash = customer.is_holding_item_of_type(/obj/item/holochip)
+	if(cash)
+		value += cash.credits
+	if((value >= the_cost) && cash)
+		return cash.spend(the_cost)
+	return FALSE //Purchase unsuccessful
+
+/**
+ * Tries to call sell_item on one of the customer's held items, if fail gives a chat message
+ *
+ * Gets both items in the customer's hands, and then tries to call sell_item on them, if both fail, he gives a chat message
+ * Arguments:
+ * * customer - (Mob REF) The mob trying to sell something
+ */
+/datum/component/trader/proc/try_sell(mob/customer)
+	if(!can_trade(customer))
+		return
+	var/mob/living/trader = parent
+	var/sold_item = FALSE
+	for(var/obj/item/an_item in customer.held_items)
+		if(sell_item(customer, an_item))
+			sold_item = TRUE
+			break
+	if(!sold_item && !can_trade(customer)) //don't talk if you are dead or fighting
+		trader.say(trader_data.return_trader_phrase(ITEM_REJECTED_PHRASE))
+
+
+/**
+ * Checks if an item is in the list of wanted items and if it is after a Yes/No radial returns generate_cash with the value of the item for the NPC
+ * Arguments:
+ * * customer - (Mob REF) The mob trying to sell something
+ * * selling - (Item REF) The item being sold
+ */
+/datum/component/trader/proc/sell_item(mob/customer, obj/item/selling)
+	var/mob/living/trader = parent
+	var/cost
+	if(!selling)
+		return FALSE
+	var/list/product_info
+	//Keep track of the typepath; rather mundane but it's required for correctly modifying the wanted_items
+	//should a product be sellable because even if it doesn't have a entry because it's a child of a parent that is present on the list
+	var/typepath_for_product_info
+	if(selling.type in wanted_items)
+		product_info = wanted_items[selling.type]
+		typepath_for_product_info = selling.type
+	else //Assume wanted_items is setup in the correct way; read wanted_items documentation for more info
+		for(var/typepath in wanted_items)
+			if(istype(selling, typepath))
+				product_info = wanted_items[typepath]
+				typepath_for_product_info = typepath
+				break
+
+	if(!product_info) //Nothing interesting to sell
+		return FALSE
+	if(product_info[TRADER_PRODUCT_INFO_QUANTITY] <= 0)
+		trader.say(trader_data.return_trader_phrase(TRADER_HAS_ENOUGH_ITEM_PHRASE))
+		return FALSE
+	cost = apply_sell_price_mods(selling, product_info[TRADER_PRODUCT_INFO_PRICE])
+	if(cost <= 0)
+		trader.say(trader_data.return_trader_phrase(ITEM_IS_WORTHLESS_PHRASE))
+		return FALSE
+	trader.say(trader_data.return_trader_phrase(INTERESTED_PHRASE))
+	trader.say("You will receive [cost] [trader_data.currency_name] for the [selling].")
+	var/list/npc_options = list(
+		"Yes" = radial_icons_cache[TRADER_RADIAL_YES],
```

Comment:
Might seem silly to make a define for `"Yes"` but we are using it as a logical key and the capitalisation and lack of punctuation might trip someone else up when touching this later